Add Daikin Madoka thermostat integration

mduran80 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add integration for Daikin BRC1H bluetooth thermostats (a.k.a madoka)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16228

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [X] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mduran80 commented:
Black and isort were stuck in a loop, each of them modifying the other changes so I had to disable isort. Any solution for this?

mduran80 commented:
@frenck can you have a look at the isort error, please?

frenck commented on .pre-commit-config.yaml:
```diff
@@ -43,6 +43,9 @@ repos:
     rev: 5.5.3
     hooks:
       - id: isort
+        args:
+          - --profile
+          - black
```
```suggestion
```

frenck commented on homeassistant/components/daikin_madoka/.gitignore:
```diff
@@ -0,0 +1,140 @@
+# Byte-compiled / optimized / DLL files
+__pycache__/
```
This file should not be here. Please remove this file.

mduran80 replied:
Done

frenck commented on homeassistant/components/daikin_madoka/sensor.py:
```diff
@@ -0,0 +1,105 @@
+# pylint: skip-file
+"""Support for Daikin Madoka HVAC sensors."""
```
Limit the addition of a new integration to a single platform. Either add climate or sensor, not both.

mduran80 replied:
Sorry for my lack of knowledge.. Does that mean one PR per platform even if they belong to the same component?

frenck replied:
Currently, this integration provides 2 platforms: `climate` and `sensor`. It should only provide 1 of those in the initial PR. More platforms can be added in subsequently PRs.

The reason is that we try to keep PRs as small as possible, in order to be able to move forward quicker.

frenck commented on homeassistant/components/daikin_madoka/translations/es.json:
```diff
@@ -0,0 +1,25 @@
+{
```
Translations are handled via Lokalise. Additional languages should not be provided. Please remove this file.

mduran80 replied:
This used to be a custom component but I decided to work a bit more and make it available to the core. Thank you for your help!

mduran80 replied:
@frenck Let me ask you a question... I would like to notify the user that the integration might need to be reloaded if the bluetooth adapter fails. Until now, in my custom component I used a logging message. What would be the correct approach here? Is it enough to use the logs?

Apart from that, do logging messages have to be lokalised? I am not sure as all the logs seem written in English even if my instance is configured in Spanish.

frenck commented on homeassistant/components/daikin_madoka/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+# pylint: skip-file
```
Don't skip files for Pylint
```suggestion
```

frenck commented on homeassistant/components/daikin_madoka/climate.py:
```diff
@@ -0,0 +1,350 @@
+# pylint: skip-file
```
```suggestion
```

frenck commented on homeassistant/components/daikin_madoka/__init__.py:
```diff
@@ -0,0 +1,114 @@
+# pylint: skip-file
```
```suggestion
```

frenck commented on homeassistant/components/daikin_madoka/const.py:
```diff
@@ -0,0 +1,14 @@
+ATTR_INSIDE_TEMPERATURE = "inside_temperature"
+ATTR_OUTSIDE_TEMPERATURE = "outside_temperature"
+SENSOR_TYPE_TEMPERATURE = "temperature"
+ATTR_STATE_ON = "on"
+ATTR_STATE_OFF = "off"
```
There are available in the Home Assistant core constants. Use those instead

frenck commented on homeassistant/components/daikin_madoka/sensor.py:
```diff
@@ -0,0 +1,105 @@
+# pylint: skip-file
```
```suggestion
```

frenck reviewed (commented):
Some comments to get you started.

Additional:
- There is an image in this PR? Please remove it.
- The Configuration Flow requires tests, with a 100% coverage

I cannot reproduce the black/isort issues you've been pointing out. The changes isort wants to make (as shown in the CI), are correct and the same in my local development environment. black doesn't trigger on the changes isort created on my end either.

Please make sure you have the correct version of black & isort.

mduran80 committed:
Add Daikin Madoka thermostat integration
Updated for lint tests
Removed climate platform and unused variables
Force CI workflow run
Clean up
Fixed discover devices wrong parameters
Linter fixes
Ensure unique id is set and early abort
Added config flow tests
Remove platform till next PR
Updated requirements
Bleak error workaround
Bleak init error workaround

mduran80 committed:
Revert precommit isort hack

mduran80 committed:
Removed core dump

mduran80 committed:
Fixed bugs and additional platform removal

mduran80 committed:
Added full testing and complete coverage

mduran80 committed:
Fixed tests

mduran80 committed:
Fixed logging

mduran80 commented:
Hey @frenck ! Could you have another look at this? I guess you must be busy with other stuff, but this is my first PR (noob) and I don't know if I have to ping you or..  just wait? Thank you!

iMicknl commented on homeassistant/components/daikin_madoka/__init__.py:
```diff
@@ -0,0 +1,100 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = vol.Schema(
```
This is a new integration and thus YAML flow can't be used, see [ADR-0010](https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md) for more context.

I see you already marked it a deprecated, what is the reason?

mduran80 replied:
I took as an example some old integration from the time both configuration ways were allowed and didn't realize I had to remove YAML support. Removing..

iMicknl commented on homeassistant/components/daikin_madoka/__init__.py:
```diff
@@ -0,0 +1,100 @@
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.typing import HomeAssistantType
+
+from . import config_flow  # noqa: F401
```
Why do you import config_flow? (and why do you disable F401)

mduran80 replied:
Legacy from when I was trying to make Flake work properly but didn't want to stop developing hehe. Removing

iMicknl commented on homeassistant/components/daikin_madoka/__init__.py:
```diff
@@ -0,0 +1,100 @@
+    for device, controller in controllers.items():
+        try:
+            await controller.start()
+        except ConnectionAbortedError as connection_aborted_error:
```
Does this mean that the integration couldn't be setup? Is this temporary or do you need to reconfigure the integration?

You might want to raise `PlatformNotReady` here, to retry setting up the integration.

mduran80 replied:
The issue here is that the integration allows you to setup several devices (specified with a comma separated list of mac addresses in a text box, I haven't found how else it could be done). I can't raise PlatformNotReady as it may be falling just for one the device and I have to let the flow continue for the rest of them. Afterwards, a message is showed for every device that is not available.

iMicknl commented on homeassistant/components/daikin_madoka/climate.py:
```diff
@@ -0,0 +1,363 @@
+    OperationModeEnum.HEAT: CURRENT_HVAC_HEAT,
+}
+
+DATA = "data"
```
Seems unused?

mduran80 replied:
Removed

iMicknl commented on homeassistant/components/daikin_madoka/climate.py:
```diff
@@ -0,0 +1,363 @@
+        """Return the temperature we try to reach."""
+
+        if self.controller.set_point.status is None:
+            return MIN_TEMP
```
Is it always set? Why set it to MIN_TEMP instead of None, when no target_temperature is set?

mduran80 replied:
Updated. Somehow it made it easier to test with the fixtures, but I replaced the values for json null and can handle empty state accordingly. Thanks!

iMicknl commented on homeassistant/components/daikin_madoka/climate.py:
```diff
@@ -0,0 +1,363 @@
+    def current_temperature(self):
+        """Return the current temperature."""
+        if self.controller.temperatures.status is None:
+            return MIN_TEMP
```
Why set it to MIN_TEMP when no temperature is known? Why not set it to None?

iMicknl commented on homeassistant/components/daikin_madoka/climate.py:
```diff
@@ -0,0 +1,363 @@
+            )
+        except ConnectionAbortedError:
+            # pylint: disable=logging-not-lazy
+            _LOGGER.info(
```
Same remark as above about severity level.

iMicknl commented on homeassistant/components/daikin_madoka/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+            # Input is valid, set data.
+            if not errors:
+
+                if user_input[CONF_DISCOVERY]:
```
nit: but would be nice to remove some of the extra whitespace/enters here

iMicknl commented on homeassistant/components/daikin_madoka/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+                step_id="user", data_schema=self.schema, errors=errors
+            )
+
+    async def async_step_import(self, user_input):
```
Not required

mduran80 replied:
Removed

iMicknl commented on homeassistant/components/daikin_madoka/manifest.json:
```diff
@@ -0,0 +1,8 @@
+  "domain": "daikin_madoka",
+  "name": "Daikin Madoka",
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/daikin",
```
This is not the documentation for the correct domain

mduran80 replied:
Lol.. copy & paste. Updated.

iMicknl commented on homeassistant/components/daikin_madoka/strings.json:
```diff
@@ -0,0 +1,25 @@
+                  "force_update": "Force disconnect from BRC1H device before start (recommended)",
+                  "discovery": "Attempt to discover the devices. They must not be connected to a mobile app"
+              },
+              "description": "Enter the Bluetooth MAC addresses of your BRC1H thermostats. Please make sure you have previously paired them with this instance as it requires human intervention:\n\n1. Forget bluetooth connection on BRC1H\n2. bluetoothctl scan on\n3. bluetoochctl scan off\n4. bluetoothctl pair [BRC1H MAC]\n5. Confirm when prompted\n6. Confirm connection on BRC1H",
```
I don't have experience with bluetooth and HA, but I wonder if this is something that should be part of the config flow. Those pairing steps.

mduran80 replied:
This can't be done from within the config flow, it requires human interaction using the OS CLI as it is handled by the OS bus.

iMicknl reviewed (commented):
I had a quick look, hopefully it is helpful :-)

iMicknl commented on homeassistant/components/daikin_madoka/__init__.py:
```diff
@@ -0,0 +1,100 @@
+    CONF_SCAN_INTERVAL,
+)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.typing import HomeAssistantType
```
Not sure if this is still the best way, or if we just use `HomeAssistant` from `from homeassistant.core` nowadays.

mduran80 replied:
Updated

mduran80 commented on homeassistant/components/daikin_madoka/climate.py:
```diff
@@ -0,0 +1,363 @@
+    def device_info(self):
+        """Return a device description for device registry."""
+
+        return self.dev_info
```
No, the properties may no be fully aligned. When I reviewed what was being done with them I just thought it would do no harm to let HA list only those that match. I will change it to the supported properties as suggested.

mduran80 commented on homeassistant/components/daikin_madoka/climate.py:
```diff
@@ -0,0 +1,363 @@
+            # pylint: disable=logging-not-lazy
+            _LOGGER.info(
+                "Could not turn on %s. "
+                + "Connection not available, please reload integration to try reenabling.",
```
You are right, this message should be in the default severity level.

mduran80 committed:
Fix climate default values when no state is available

mduran80 committed:
Remove YAML support and unused code

mduran80 committed:
Change default logging severity level for messages to the user

mduran80 committed:
Improve device info to use only allowed properties and fix manifest

mduran80 committed:
Fixed device info and status

mduran80 committed:
Update config_flow.py

mduran80 committed:
Bumped pymadoka version to prevent DBUS connection errors

mduran80 committed:
Update manifest.json
